release notes for the 1.15 release (#19584)
apache/flink: docs/content/release-notes/flink-1.15.md

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Release notes - Flink 1.15

These release notes discuss important aspects, such as configuration, behavior, 
or dependencies, that changed between Flink 1.14 and Flink 1.15. Please read these 
notes carefully if you are planning to upgrade your Flink version to 1.15.

## Summary of changed dependency names

There are three changes in Flink 1.15 that require updating dependency names when 
upgrading from earlier versions:

* The newly introduced module `flink-table-planner-loader` (FLINK-25128) could 
  replace the legacy `flink-table-planner_2.12`. As a consequence, `flink-table-uber` 
  has been split into `flink-table-api-java-uber`, `flink-table-planner(-loader), 
  and `table-runtime`. Besides, the artifactId of `flink-sql-client` has no Scala 
  suffix (_2.11 / _2.12) anymore, and Scala users need to explicitly add a 

>   dependency to `flink-table-api-scala` or `flink-table-api-scala-bridge`.

Reviewer: This isn't phrased in a user-friendly way. We have to come up with something along the lines of "if you used A, now use B".
IMO we also shouldn't combine so many things into 1 item (split/suffix removal/scala opt-in).

Reviewer 2: Thanks @zentol for the review! I'll split them into different items and give a summary table for the relationship.

  suffix (\_2.11 / \_2.12) any more.
* The FileSystem connector is no longer a part of the `flink-table-uber` module 
  and changed to an optional dedicated `flink-connector-files` module 
  (FLINK-24687). Besides, the artifactId of `flink-orc`, `flink-orc-nohive`, 
  `flink-parquet` has no Scala suffix (\_2.11 / \_2.12) anymore.

## JDK Upgrade

The support of Java 8 is now deprecated and will be removed in a future release 
([FLINK-25247](https://issues.apache.org/jira/browse/FLINK-25247)). We recommend 
all users to migrate to Java 11.

The default Java version in the Flink docker images is now Java 11 
([FLINK-25251](https://issues.apache.org/jira/browse/FLINK-25251)). 
There are images built with Java 8, tagged with “java8”. 

## DataStream API

### [TypeSerializer version mismatch during eagerly restore (FLINK-24858)](https://issues.apache.org/jira/browse/FLINK-24858)

This ticket resolves an issue that during state migration between Flink versions
the wrong serializer might have been picked.

When upgrading from Flink 1.13.x please immediately choose 1.14.3 or higher and 
skip 1.14.0, 1.14.1, 1.14.2 because all are affected and it might prevent your 

> job from starting.

Reviewer: Should be able to remove this given that we're talking about 1.15.

Table functions that are called using Scala implicit conversions have been updated 
to use the new type system and new type inference. Users are requested to update 
their UDFs or use the deprecated `TableEnvironment.registerFunction` to restore 
the old behavior temporarily by calling the function via name.

### [Propagate executor config to TableConfig (FLINK-26421)](https://issues.apache.org/jira/browse/FLINK-26421)

`flink-conf.yaml` and other configurations from outer layers (e.g. CLI) are now 
propagated into `TableConfig`. Even though configuration set directly in `TableConfig` 
has still precedence, this change can have side effects if table configuration 
was accidentally set in other layers.

### [Remove pre FLIP-84 methods (FLINK-26090)](https://issues.apache.org/jira/browse/FLINK-26090)

The previously deprecated methods `TableEnvironment.execute`, `Table.insertInto`,
`TableEnvironment.fromTableSource`, `TableEnvironment.sqlUpdate`, and 
`TableEnvironment.explain` have been removed. Please use the provided 
alternatives introduced in 

> [FLIP-84](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=134745878).

Reviewer: IMO we shouldn't link to FLIPs in user-facing content. If we want to refer them to anything, then I guess it should be the javadocs that should contain migration instructions.

Author: changed the text, will also add some javadocs links.

### [Fix parser generator warnings (FLINK-26053)](https://issues.apache.org/jira/browse/FLINK-26053)

`STATEMENT` is a reserved keyword now. Use backticks to escape tables, fields and 
other references.

### [Expose uid generator for DataStream/Transformation providers (FLINK-25990)](https://issues.apache.org/jira/browse/FLINK-25990)

`DataStreamScanProvider` and `DataStreamSinkProvider` for table connectors received 
an additional method that might break implementations that used lambdas before. 
We recommend static classes as a replacement and future robustness.

### [Add new STATEMENT SET syntax (FLINK-25392)](https://issues.apache.org/jira/browse/FLINK-25392)

It is recommended to update statement sets to the new SQL syntax: 

```SQL
EXECUTE STATEMENT SET BEGIN ... END;
EXPLAIN STATEMENT SET BEGIN ... END;
```

> ### [Check & possible fix decimal precision and scale for all Aggregate functions (FLINK-24809)](https://issues.apache.org/jira/browse/FLINK-functions)

Reviewer: broken jira link

1.14.0 and 1.14.1. It restores the behavior of 1.13 to be consistent with Hive/Spark.

### [Clarify semantics of DecodingFormat and its data type (FLINK-24776)](https://issues.apache.org/jira/browse/FLINK-24776)

The `DecodingFormat` interface was used for both projectable and non-projectable
formats which led to inconsistent implementations. The `FileSystemTableSource` 
has been updated to distinguish between those two interfaces now. Users that 
implement custom formats for `FileSystemTableSource` might need to verify the 
implementation and make sure to implement `ProjectableDecodingFormat` if necessary.

### [Update user document for batch window tvf (FLINK-24760)](https://issues.apache.org/jira/browse/FLINK-24760)

All window table-valued functions can now also be used in batch mode.

### [Add user document for Window Deduplication (FLINK-24656)](https://issues.apache.org/jira/browse/FLINK-24656)

Flink SQL now also supports Window Deduplication as a new window table-valued function.
`WindowDeduplication` is a deduplication operator which removes rows that duplicate
over a set of columns, keeping only the first one or the last one for each window

> and other partitioned keys.

Reviewer: I don't think these 2 belong in here. They are ultimately new features, and not a cause for issues during upgrades.

the behavior of 1.13 to be consistent with Hive/Spark.

### [Use the new casting rules in TableResult#print (FLINK-24685)](https://issues.apache.org/jira/browse/FLINK-24685)

The string representation of `BOOLEAN` columns from DDL results 
(`true/false -> TRUE/FALSE`), and row columns in DQL results 
(`+I[...] -> (...)`) has changed for printing.

### [Sinks built with the unified sink framework do not receive timestamps when used in Table API (FLINK-24608)](https://issues.apache.org/jira/browse/FLINK-Table)

This adds an additional operator to the topology if the new sink interfaces are used 
(e.g. for Kafka). It could cause issues in 1.14.1 when restoring from a 1.14 savepoint. 
A workaround is to cast the time attribute to a regular timestamp in the SQL statement
closely before the sink.

### [SQL functions should return `STRING` instead of `VARCHAR(2000)` (FLINK-24586)](https://issues.apache.org/jira/browse/FLINK-24586)

Functions that returned `VARCHAR(2000)` in 1.14, return `VARCHAR` with maximum 
length now. In particular this includes:

```SQL
SON_VALUE
CHR
REVERSE
SPLIT_INDEX
REGEXP_EXTRACT
PARSE_URL
FROM_UNIXTIME
DECODE
DATE_FORMAT
CONVERT_TZ
```

## Connectors

> ### [(Introducing GCS FileSystem Support and) Update GCS documentation (FLINK-25577)](https://issues.apache.org/jira/browse/FLINK-25577)

Reviewer: also not relevant for upgrades imo

In particular, this plugin supports the `RecoverableWriter` interface, which 
allows it to be used with file sinks.

Under the hood, the `flink-gs-fs-hadoop` uses Google's `gcs-connector` Hadoop 
library for basic read/write operations, and it uses Google's `google-cloud-storage` 
library to implement `RecoverableWriter` functionality.

### [Remove MapR filesystem (FLINK-25553)](https://issues.apache.org/jira/browse/FLINK-25553)

Support for the MapR FileSystem has been dropped.

### [Make flink-connector-base dependency usage consistent across all connectors (FLINK-25927)](https://issues.apache.org/jira/browse/FLINK-25927)

`flink-connector-base` was previously inconsistently used in connectors (directly 
shaded in some and transitively pulled in via `flink-connector-files` which was 
itself shaded in the table uber jar). 
[FLINK-24687](https://issues.apache.org/jira/browse/FLINK-24687) moved 
`flink-connector-files` 
out of the `flink-table` uber jar. In order to address this, Flink 1.15 implements
a combined approach for both making usage of `flink-connector-base` consistent 
and ensuring a smooth transition for both Flink users and for external connector
developers:

> 1. all internal Flink connectors that depend on `flink-connector-base` now shade

Reviewer: this was actually partially reverted; cc @AHeise @afedulov

Reviewer 3: Yes, the commit with the associated release note was reverted.

Reviewer 3: Some background: unfortunately some classes within the `flink-connector-base`,  such as `DeliveryGuarantee`, that are public and can be used in the user code and need to be first moved into `flink-core`. The same applies to the `HybridSource`.

Reviewer 2: Thanks @afedulov for the information! I'll then first remove this issue from the release note.

   and relocate it.
2. for compatibility, until external developers implement the same change,
   flink-connector-base is also included into `flink-dist`.
3. It is recommended for the external connector developers to shade and relocate
   `flink-connector-base` in their implementations.

### [Merge flink-connector-testing into flink-connector-test-utils (FLINK-25712)](https://issues.apache.org/jira/browse/FLINK-25712)

> The `flink-connector-testing` module has been removed and should use
> `flink-connector-test-utils` module instead.

Reviewer: ```suggestion
The `flink-connector-testing` module has been removed and users should use
`flink-connector-test-utils` module instead.
```

> ### [Support to disable caching missing key for lookup cache (FLINK-24861)](https://issues.apache.org/jira/browse/FLINK-24861)

Reviewer: not relevant for these release notes

You can now configure the JDBC connector to enable or disable caching results 
for keys which don't exist in the database. This is enabled by default.

### [Support partition keys through metadata (for FileSystem connector) (FLINK-24617)](https://issues.apache.org/jira/browse/FLINK-24617)

Now the formats implementing `BulkWriterFormatFactory` don't need to implement
partition keys reading anymore, as it's managed internally by `FileSystemTableSource`.

## Runtime & Coordination

### [Implement file-based JobResultStore (FLINK-25431)](https://issues.apache.org/jira/browse/FLINK-25431)

Introduces a file-based version of the `JobResultStore` with new configuration

> parameters `job-result-store.storage-path` and `job-result-store.delete-on-commit`. 

Reviewer 4: ```suggestion
parameters [job-result-store.storage-path](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#job-result-store-storage-path) and [job-result-store.delete-on-commit](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#job-result-store-delete-on-commit). 
```

This feature persists the final state of a job including its state of cleanup.

### [Integrate retry strategy for cleanup stage (FLINK-25433)](https://issues.apache.org/jira/browse/FLINK-25433)

Adds retry logic to the cleanup steps of a finished job. This feature changes the
way Flink jobs are cleaned up. Instead of trying once to clean up the job, this
step will be repeated until it succeeds. Users are meant to fix the issue that
prevents Flink from finalizing the job cleanup. The retry functionality can be
configured and disabled. More details can be found 
[in the documentation](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#retryable-cleanup).

### [Introduce common interfaces for cleaning up local and global job data (FLINK-25432)](https://issues.apache.org/jira/browse/FLINK-25432)

Utilizes the newly added component `JobResultStore` (see 
[FLINK-25431](https://issues.apache.org/jira/browse/FLINK-25431)) to pick up

> already finished but not properly cleaned up jobs after a failover.

Reviewer: This is an internal thing that doesn't belong into the release notes afaik. @XComp

Reviewer 4: You're actually right. The important things are the retry configuration (FLINK-25433) and the file-based JRS (FLINK-25431). I will remove the release notes from FLINK-25432.

Reviewer 4: Done... ✅

### [Introduce explicit shutdown signaling between `TaskManager` and `JobManager` (FLINK-25277)](https://issues.apache.org/jira/browse/FLINK-25277)

Reduce down-scaling delay in reactive mode through signaling between `TaskManager` 
and `JobManager`.

### [FLIP-187: Adaptive Batch Scheduler (FLINK-24892)](https://issues.apache.org/jira/browse/FLINK-24892)

In 1.15, we introduced a new scheduler to Flink: Adaptive Batch Scheduler. The 
new scheduler can automatically decide parallelisms of job vertices for batch
jobs, according to the size of data volume each vertex needs to process.

More information about this feature and how to enable it could be found in 

> [the documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/elastic_scaling/#adaptive-batch-scheduler.)

Reviewer: new feature; should be reserved to blog post

### [Release `TaskManagerJobMetricGroup` with the last slot rather than task (FLINK-24864)](https://issues.apache.org/jira/browse/FLINK-24864)

TM-per-job metrics are now released upon last slot release rather than task, i.e. 
may stay on TM longer.

> ### [Issue in exceptions API(/jobs/:jobid/exceptions) in flink 1.13.2 (FLINK-25096)](https://issues.apache.org/jira/browse/FLINK-25096)

Reviewer: @XComp please change the jira title to be more descriptive

Reviewer 4: Good point - I updated the Jira issue title. Sorry for not thinking about it while losing the issue.

Reviewer 4: Done ✅

is as follows:

* Any dependency to one of the following modules needs to be updated to no longer include a suffix:
  
    ```
    flink-cep
    flink-clients
    flink-connector-elasticsearch-base
    flink-connector-elasticsearch6
    flink-connector-elasticsearch7
    flink-connector-gcp-pubsub
    flink-connector-hbase-1.4
    flink-connector-hbase-2.2
    flink-connector-hbase-base
    flink-connector-jdbc
    flink-connector-kafka
    flink-connector-kinesis
    flink-connector-nifi
    flink-connector-pulsar
    flink-connector-rabbitmq
    flink-container
    flink-dstl-dfs
    flink-gelly
    flink-hadoop-bulk
    flink-kubernetes
    flink-runtime-web
    flink-sql-connector-elasticsearch6
    flink-sql-connector-elasticsearch7
    flink-sql-connector-hbase-1.4
    flink-sql-connector-hbase-2.2
    flink-sql-connector-kafka
    flink-sql-connector-kinesis
    flink-sql-connector-rabbitmq
    flink-state-processor-api
    flink-statebackend-rocksdb
    flink-streaming-java
    flink-test-utils
    flink-yarn
    flink-table-api-java-bridge
    flink-table-runtime
    flink-sql-client
    flink-orc
    flink-orc-nohive
    flink-parquet
    ```

> * For Table / SQL users, the new module `flink-table-planner-loader` replaces `flink-table-planner_2.12`
>   and avoids the need for a Scala suffix. As a consequence, `flink-table-uber` has been split into `flink-table-api-java-uber`,
>   `flink-table-planner(-loader)`, and `flink-table-runtime`. Scala users need to explicitly add a dependency
>   to `flink-table-api-scala` or `flink-table-api-scala-bridge`. For backwards compatibility, users can still
>   swap it with `flink-table-planner_2.12` located in `opt/`.

Reviewer: ```suggestion
* For Table / SQL users, the new module `flink-table-planner-loader` replaces `flink-table-planner_2.12`
  and avoids the need for a Scala suffix. For backwards compatibility, users can still
  swap it with `flink-table-planner_2.12` located in `opt/`.
  `flink-table-uber` has been split into `flink-table-api-java-uber`,
  `flink-table-planner(-loader)`, and `flink-table-runtime`. Scala users need to explicitly add a dependency
  to `flink-table-api-scala` or `flink-table-api-scala-bridge`.
```

`flink-connector-files` JAR in the `/lib` directory of a Flink distribution.

## JDK Upgrade

The support of Java 8 is now deprecated and will be removed in a future release 
([FLINK-25247](https://issues.apache.org/jira/browse/FLINK-25247)). We recommend 
all users to migrate to Java 11.

The default Java version in the Flink docker images is now Java 11 
([FLINK-25251](https://issues.apache.org/jira/browse/FLINK-25251)). 
There are images built with Java 8, tagged with “java8”. 

## Drop support for Scala 2.11

Support for Scala 2.11 has been removed in
[FLINK-20845](https://issues.apache.org/jira/browse/FLINK-20845). 
All Flink dependencies that (transitively)
depend on Scala are suffixed with the Scala version that they are built for, for
example `flink-streaming-scala_2.12`. Users should update all Flink dependecies,
changing "2.11" to "2.12".

Scala versions (2.11, 2.12, etc.) are not binary compatible with one another. That
also means that there's no guarantee that you can restore from a savepoint, made
with a Flink Scala 2.11 application, if you're upgrading to a Flink Scala 2.12
application. This depends on the data types that you have been using in your
application.

> Also, the Scala Shell/REPL has been removed in
> [FLINK-24360](https://issues.apache.org/jira/browse/FLINK-24360).

Reviewer: ```suggestion
The Scala Shell/REPL has been removed in
[FLINK-24360](https://issues.apache.org/jira/browse/FLINK-24360).
```

## DataStream API

#### TypeSerializer version mismatch during eagerly restore 

##### [FLINK-24858](https://issues.apache.org/jira/browse/FLINK-24858)

This ticket resolves an issue that during state migration between Flink versions

> the wrong serializer might have been picked.

Reviewer: Is this relevant for the upgrade? Sounds more like a plain bug fix that users don't need to be aware of.

Reviewer 2: Yes, it should be a plain bug fix. I'll then remove it.

## Table API & SQL

> #### Make the legacy behavior disabled by default 

Reviewer: ```suggestion
#### Disable the legacy casting behavior by default 
```

##### [FLINK-26551](https://issues.apache.org/jira/browse/FLINK-26551)

The legacy casting behavior has been disabled by default. This might have 
implications on corner cases (string parsing, numeric overflows, to string 
representation, varchar/binary precisions). Set 

> `table.exec.legacy-cast-behaviour=ENABLED` to restore the old behavior.

Reviewer: Are there plans to remove the legacy-cast-behavior (i.e., is it deprecated)?

Reviewer 2: Checking with the SQL guys the legacy behavior should not be deprecated in short period since the difference might be very large. It might be removed in, for example, 2.0

`TableResult`, `ResultKind`, `StatementSet` and `ExplainDetail`.

#### Fix parser generator warnings 

##### [FLINK-26053](https://issues.apache.org/jira/browse/FLINK-26053)

`STATEMENT` is a reserved keyword now. Use backticks to escape tables, fields and 
other references.

#### Expose uid generator for DataStream/Transformation providers 

##### [FLINK-25990](https://issues.apache.org/jira/browse/FLINK-25990)

`DataStreamScanProvider` and `DataStreamSinkProvider` for table connectors received 
an additional method that might break implementations that used lambdas before. 
We recommend static classes as a replacement and future robustness.

#### Add new STATEMENT SET syntax 

##### [FLINK-25392](https://issues.apache.org/jira/browse/FLINK-25392)

It is recommended to update statement sets to the new SQL syntax: 

```SQL
EXECUTE STATEMENT SET BEGIN ... END;
EXPLAIN STATEMENT SET BEGIN ... END;
```

#### Check & possible fix decimal precision and scale for all Aggregate functions 

##### [FLINK-24809](https://issues.apache.org/jira/browse/FLINK-24809)

This changes the result of a decimal `SUM()` with retraction and `AVG()` between

> 1.14.0 and 1.14.1. It restores the behavior of 1.13 to be consistent with Hive/Spark.

Reviewer: So...if you're upgrading from 1.13.X/1.14.1+ nothing changes?

Reviewer 2: The issue itself is merged to 1.14 / 1.13, thus I think for high minor version of 1.13.x it should be such case. I'll remove the part irrelevant to 1.15.

Reviewer 2: By checking with the SQL guys, the story would be like 
- If we mark 1.13 as the base
- In 1.14, we do some modification, makes it partly compatible with Hive/spark and partly compatible with sqlserver
- In 1.15, we revert the part that compatible with sqlserver back to the behavior of 1.13, makes it compatible with Hive/Spark as a whole

I'll try modifying the release notes.

##### [FLINK-24446](https://issues.apache.org/jira/browse/FLINK-24446)

`STRING` to `TIMESTAMP(_LTZ)` casting now considers fractional seconds.
Previously fractional seconds of any precision were ignored.

#### Sinks built with the unified sink framework do not receive timestamps when used in Table API 

##### [FLINK-24608](https://issues.apache.org/jira/browse/FLINK-Table)

This adds an additional operator to the topology if the new sink interfaces are used 
(e.g. for Kafka). It could cause issues in 1.14.1 when restoring from a 1.14 savepoint. 
A workaround is to cast the time attribute to a regular timestamp in the SQL statement
closely before the sink.

#### SQL functions should return `STRING` instead of `VARCHAR(2000)` 

##### [FLINK-24586](https://issues.apache.org/jira/browse/FLINK-24586)

Functions that returned `VARCHAR(2000)` in 1.14, return `VARCHAR` with maximum 
length now. In particular this includes:

```SQL
SON_VALUE
CHR
REVERSE
SPLIT_INDEX
REGEXP_EXTRACT
PARSE_URL
FROM_UNIXTIME
DECODE
DATE_FORMAT
CONVERT_TZ
```

#### Support IS JSON for Table API 

##### [FLINK-16501](https://issues.apache.org/jira/browse/FLINK-16501)

> `IS JSON` does not return `NULL` anymore but always `FALSE` (even if the argument is `NULL`).

Reviewer: the ticket title and description don't really fit. The title makes it seem sound like a new feature, while the description seems more like a behavior change.

The results of `Table#print` have changed to be closer to actual SQL data types.
E.g. decimal is printing correctly with leading/trailing zeros.

## Connectors

#### Remove MapR filesystem 

##### [FLINK-25553](https://issues.apache.org/jira/browse/FLINK-25553)

Support for the MapR FileSystem has been dropped.

#### Merge flink-connector-testing into flink-connector-test-utils 

##### [FLINK-25712](https://issues.apache.org/jira/browse/FLINK-25712)

The `flink-connector-testing` module has been removed and users should use
`flink-connector-test-utils` module instead.

#### Support partition keys through metadata (for FileSystem connector) 

##### [FLINK-24617](https://issues.apache.org/jira/browse/FLINK-24617)

Now the formats implementing `BulkWriterFormatFactory` don't need to implement
partition keys reading anymore, as it's managed internally by `FileSystemTableSource`.

#### Port ElasticSearch Sink to new Unified Sink API (FLIP-143) 

##### [FLINK-24323](https://issues.apache.org/jira/browse/FLINK-24323)

`ElasticsearchXSinkBuilder` supersedes `ElasticsearchSink.Builder` and provides at-least-once writing with the

> new unified sink interface supporting both batch and streaming mode of DataStream API.To upgrade, please stop with savepoint.

Reviewer 2: I also think we could remove this statement.

(`org.apache.flink.streaming.connectors.elasticsearch7.ElasticsearchSink`)
and depend on their own elasticsearch-rest-high-level-client version,
updating the client dependency to a version >= 7.14.0 is required due to internal changes.

#### Reduce legacy in Table API connectors 

##### [FLINK-24397](https://issues.apache.org/jira/browse/FLINK-24397)

The old JDBC connector (indicated by `connector.type=jdbc` in DDL) has been removed.
If not done already, users need to upgrade to the newer stack (indicated by `connector=jdbc` in DDL).

## Runtime & Coordination

#### Integrate retry strategy for cleanup stage 

##### [FLINK-25433](https://issues.apache.org/jira/browse/FLINK-25433)

Adds retry logic to the cleanup steps of a finished job. This feature changes the
way Flink jobs are cleaned up. Instead of trying once to clean up the job, this
step will be repeated until it succeeds. Users are meant to fix the issue that
prevents Flink from finalizing the job cleanup. The retry functionality can be
configured and disabled. More details can be found 
[in the documentation](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#retryable-cleanup).

> 
> #### Introduce common interfaces for cleaning up local and global job data 
> 
> ##### [FLINK-25432](https://issues.apache.org/jira/browse/FLINK-25432)
> 
> Utilizes the newly added component `JobResultStore` (see 
> [FLINK-25431](https://issues.apache.org/jira/browse/FLINK-25431)) to pick up
> already finished but not properly cleaned up jobs after a failover.

Reviewer 2: ```suggestion
```

#### Introduce explicit shutdown signaling between `TaskManager` and `JobManager` 

##### [FLINK-25277](https://issues.apache.org/jira/browse/FLINK-25277)

Reduce down-scaling delay in reactive mode through signaling between `TaskManager` 
and `JobManager`.

#### Release `TaskManagerJobMetricGroup` with the last slot rather than task 

##### [FLINK-24864](https://issues.apache.org/jira/browse/FLINK-24864)

TM-per-job metrics are now released upon last slot release rather than task, i.e. 

> may stay on TM longer.

Reviewer: This was meant to be added as a suggestion :/

> TM-per-job metrics are now released upon last slot release rather than task, i.e. 
> may stay on TM longer.

Reviewer 2: ```suggestion
Job metrics on the TaskManager are now removed when the last slot is released, rather than the last task.
This means they may be reported for a longer time than before and when no tasks are running on the TaskManager.
```

> may stay on TM longer.

Reviewer 2: I'll modify this part accordingly~

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Release notes - Flink 1.15

These release notes discuss important aspects, such as configuration, behavior, 
or dependencies, that changed between Flink 1.14 and Flink 1.15. Please read these 
notes carefully if you are planning to upgrade your Flink version to 1.15.

## Summary of changed dependency names

There are Several changes in Flink 1.15 that require updating dependency names when 
upgrading from earlier versions, mainly including the effort to opting-out Scala dependencies
from non-scala modules and reorganize table modules. A quick checklist of the dependency changes
is as follows:

* Any dependency to one of the following modules needs to be updated to no longer include a suffix:
  
    ```
    flink-cep
    flink-clients
    flink-connector-elasticsearch-base

>     flink-connector-elasticsearch5

Reviewer: ```suggestion
```
Not part of 1.15.

    flink-connector-elasticsearch6
    flink-connector-elasticsearch7
    flink-connector-gcp-pubsub
    flink-connector-hbase-1.4
    flink-connector-hbase-2.2
    flink-connector-hbase-base
    flink-connector-jdbc
    flink-connector-kafka
    flink-connector-kinesis
    flink-connector-nifi
    flink-connector-pulsar
    flink-connector-rabbitmq
    flink-connector-testing

>     flink-connector-twitter
>     flink-connector-wikiedits

Reviewer: ```suggestion
```
Not part of 1.15.

>     flink-connector-testing
>     flink-connector-twitter
>     flink-connector-wikiedits

Reviewer: ```suggestion
```
Not part of 1.15.

    flink-container
    flink-dstl-dfs
    flink-gelly
    flink-hadoop-bulk
    flink-kubernetes
    flink-runtime-web
    flink-sql-connector-elasticsearch6
    flink-sql-connector-elasticsearch7
    flink-sql-connector-hbase-1.4
    flink-sql-connector-hbase-2.2
    flink-sql-connector-kafka
    flink-sql-connector-kinesis
    flink-sql-connector-rabbitmq
    flink-state-processor-api
    flink-statebackend-rocksdb
    flink-streaming-java
    flink-table-api-java-bridge
    flink-test-utils
    flink-yarn

>     flink-table-uber

Reviewer 2: Also cc @zentol @twalthr for a double check for this part~

>     flink-cep
>     flink-clients
>     flink-connector-elasticsearch-base
>     flink-connector-elasticsearch5
>     flink-connector-elasticsearch6
>     flink-connector-elasticsearch7
>     flink-connector-gcp-pubsub
>     flink-connector-hbase-1.4
>     flink-connector-hbase-2.2
>     flink-connector-hbase-base
>     flink-connector-jdbc
>     flink-connector-kafka
>     flink-connector-kinesis
>     flink-connector-nifi
>     flink-connector-pulsar
>     flink-connector-rabbitmq
>     flink-connector-testing
>     flink-connector-twitter
>     flink-connector-wikiedits
>     flink-container
>     flink-dstl-dfs
>     flink-gelly
>     flink-hadoop-bulk
>     flink-kubernetes
>     flink-runtime-web
>     flink-sql-connector-elasticsearch6
>     flink-sql-connector-elasticsearch7
>     flink-sql-connector-hbase-1.4
>     flink-sql-connector-hbase-2.2
>     flink-sql-connector-kafka
>     flink-sql-connector-kinesis
>     flink-sql-connector-rabbitmq
>     flink-state-processor-api
>     flink-statebackend-rocksdb
>     flink-streaming-java
>     flink-table-api-java-bridge
>     flink-test-utils
>     flink-yarn
>     flink-table-uber
>     flink-table-runtime
>     flink-orc
>     flink-orc-nohive
>     flink-parquet

Reviewer: ```suggestion
    flink-cep
    flink-clients
    flink-connector-elasticsearch-base
    flink-connector-elasticsearch6
    flink-connector-elasticsearch7
    flink-connector-gcp-pubsub
    flink-connector-hbase-1.4
    flink-connector-hbase-2.2
    flink-connector-hbase-base
    flink-connector-jdbc
    flink-connector-kafka
    flink-connector-kinesis
    flink-connector-nifi
    flink-connector-pulsar
    flink-connector-rabbitmq
    flink-container
    flink-dstl-dfs
    flink-gelly
    flink-hadoop-bulk
    flink-kubernetes
    flink-runtime-web
    flink-sql-connector-elasticsearch6
    flink-sql-connector-elasticsearch7
    flink-sql-connector-hbase-1.4
    flink-sql-connector-hbase-2.2
    flink-sql-connector-kafka
    flink-sql-connector-kinesis
    flink-sql-connector-rabbitmq
    flink-state-processor-api
    flink-statebackend-rocksdb
    flink-streaming-java
    flink-table-api-java-bridge
    flink-test-utils
    flink-yarn
    flink-table-uber
    flink-table-runtime
    flink-orc
    flink-orc-nohive
    flink-parquet
```
These don't exist in 1.15 anymore.

  For backwards compatibility, users can still swap it with `flink-table-planner_2.12` located in `opt/`.

The detail of the involved issues are listed as follows.

#### Add support for opting-out of Scala 

##### [FLINK-20845](https://issues.apache.org/jira/browse/FLINK-20845)

The Java DataSet/-Stream APIs are now independent of Scala and no longer transitively depend on it.

The implications are the following:

* If you only intend to use the Java APIs, with Java types,
then you can opt-in to a Scala-free Flink by removing the `flink-scala` jar from the `lib/` directory of the distribution.
You are then free to use any Scala version and Scala libraries.
You can either bundle Scala itself in your user-jar; or put into the `lib/` directory of the distribution.

* If you relied on the Scala APIs, without an explicit dependency on them,
  then you may experience issues when building your projects. You can solve this by adding explicit dependencies to
  the APIs that you are using. This should primarily affect users of the Scala `DataStream/CEP` APIs.

* A lot of modules have lost their Scala suffix. 

>   Any dependency on one of the modules listed below needs to be updated. 

Reviewer: ```suggestion
```
